### PR TITLE
Add a system tray icon

### LINKED ISSUE #120: Minimize Mirage to system tray
Currently, mirage is unable to minimize to system tray. This clutters the desktop when mirage is not immediately needed. In addition, this feature is standard among most chat clients.

### PR DESCRIPTION
**Pull request by vSLG**
These commits will add a system tray icon and a lock file in the current configuration.

Make use of lock files in configuration directory for preventing multiple instances of Mirage from being running at the same, at least for a single configuration. Running multiple instances of Mirage with the same configuration is known to cause encryption errors in one or more instances.

When attempting to open another Mirage instance with the same configuration, the duplicated process will create a file named `.show` in the configuration directory and exit. The main Mirage instance has a file watcher for the `.show` file, and when it detects it, it brings focus to currently running Mirage.

Options of the system tray icon:
- Hide/show Mirage
- Quit Mirage

Closes #120

### CODE REVIEW

File: src/gui/Window.qml [hunk 1]
```diff
@@ -100,4 +113,51 @@ ApplicationWindow {
...
+    SystemTrayIcon {
+        property string iconPack: theme ? theme.icons.preferredPack : "thin"
+
+        visible: true
+        tooltip: qsTr("Mirage")
+        icon.source: `../icons/${iconPack}/tray-icon.png`
+
+        onActivated:
+            if (reason !== SystemTrayIcon.Context)
+                window.drawAttention()
```
Comment (mirukana) [inline src/gui/Window.qml:126]: It would be useful to hide/show on left click (instead of only showing) and close on middle click, that way all the menu operations are doable in one click

File: src/gui/Window.qml [hunk 2]
```diff
@@ -100,4 +113,51 @@ ApplicationWindow {
...
+
+
+        menu: Menu {
+            MenuItem {
+                text: window.visible ? "Hide Mirage" : "Show Mirage"
```
Comment (mirukana) [inline src/gui/Window.qml:131]: `"Open " + Qt.application.displayName` and `"Minimize to tray"` would be better for texts

File: src/gui/Window.qml [hunk 3]
```diff
@@ -100,4 +113,51 @@ ApplicationWindow {
...
+                    window.visible ?
+                    window.hide() :
+                    window.drawAttention()
+            }
+
+            MenuItem {
+                text: qsTr("Quit Mirage")
+                onTriggered: {
+                    Qt.quit()
+                }
```
Comment (mirukana) [inline src/gui/Window.qml:142]: This can fit in one line without the braces

File: src/main.cpp [hunk 4]
```diff
@@ -86,6 +93,37 @@ void onExitSignal(int signum) {
 }
 
 
+bool setLockFile(QString configPath) {
+
+    QDir configDir(configPath);
+
+    if (! configDir.mkpath(".")) {
+        qFatal("Could not create config file.");
```
Comment (mirukana) [inline src/main.cpp:101]: Directory, not file
Also, don't end single sentences with a period, to match the style used by the UI and other log lines (applies to other places where you logged too).

File: src/main.cpp [hunk 5]
```diff
@@ -86,6 +93,37 @@ void onExitSignal(int signum) {
...
+    }
+
+    lockFile = new QLockFile(configDir.absoluteFilePath(".lock"));
+    lockFile->tryLock(0);
+
+    switch (lockFile->error()) {
+        case QLockFile::NoError: {
+            return true;
+        }
+        case QLockFile::LockFailedError: {
```
Comment (mirukana) [inline src/main.cpp:112]: No need for the `case` braces, unless you're declaring variables inside a `case`

File: src/main.cpp [hunk 6]
```diff
@@ -86,6 +93,37 @@ void onExitSignal(int signum) {
...
+            qWarning("Opening already running Mirage instance.");
```
Comment (mirukana) [inline src/main.cpp:113]: Just `"Opening already running instance"` will do

File: src/main.cpp [hunk 7]
```diff
@@ -95,6 +133,16 @@ int main(int argc, char *argv[]) {
     QApplication::setApplicationDisplayName("Mirage");
     QApplication::setApplicationVersion("0.6.2");
     QApplication::setAttribute(Qt::AA_EnableHighDpiScaling);
+
+    QString customConfigDir(qEnvironmentVariable("MIRAGE_CONFIG_DIR"));
+    QString configDir(
+        customConfigDir.isEmpty() ?
+        QStandardPaths::writableLocation(QStandardPaths::GenericConfigLocation)
+        + "/" + QCoreApplication::applicationName() :
```
Comment (mirukana) [inline src/main.cpp:141]: Use `QApplication::` since that's what we create in the main function

File: src/main.cpp [hunk 8]
```diff
@@ -135,6 +183,9 @@ int main(int argc, char *argv[]) {
     QQmlEngine engine;
     QQmlContext *objectContext = new QQmlContext(engine.rootContext());
 
+    // For being able to use Qt.quit() in QML side
```
Comment (mirukana) [inline src/main.cpp:186]: "To be able to use `Qt.quit()` from QML" is a more correct way to say this

File: src/gui/Window.qml [hunk 9]
```diff
@@ -22,6 +24,7 @@ ApplicationWindow {
     property var uiState: ({})
     property var history: ({})
     property var theme: null
+    property string configDir
```
Comment (mirukana) [inline src/gui/Window.qml:27]: Avoid using abbreviations in variable/functions/etc names (Qt coding conventions), `settingsFolder` would be better (applies anyway you've written "configDir")

File: src/gui/Window.qml [hunk 10]
```diff
@@ -64,6 +67,11 @@ ApplicationWindow {
         )
     }
 
+    function drawAttention() {
+        window.show()
+        window.raise()
+        window.requestActivate()
```
Comment (mirukana) [inline src/gui/Window.qml:73]: `drawAttention` is a confusing name, `restoreFromTray` would be more fitting (as oppose to "minimize to tray").
Are all three `window.` lines really needed? `requestActivate()` works directly on my desktop, not sure if this is the case elsewhere.

**Review by mirukana — CHANGES_REQUESTED** [review]
Everything seems functional, great work! The rest to fix is mostly just code style.

File: src/gui/Window.qml [hunk 11]
```diff
@@ -80,6 +88,11 @@ ApplicationWindow {
     onUiStateChanged: py.saveConfig("ui_state", uiState)
     onHistoryChanged: py.saveConfig("history", history)
 
+    onClosing: {
+        close.accepted = ! settings.closeMinimizesToTray
+        onTriggered: if (settings.closeMinimizesToTray) hide()
```
[the remark this answers is not in the document]
Comment (vSLG) [reply src/gui/Window.qml]: We also need a `Qt.quit()` on an else, because KDE Plasma will minimize to tray by default when clicking on X button.

File: src/gui/Window.qml (hunk 10 above)
Comment (vSLG) [reply src/gui/Window.qml:71]: Maybe different DEs/WMs have different behavior on each call. On KDE Plasma, without `show()`, it will not bring up anything. `raise()` in here brings up the window on top of everything. Without `raise()`, Mirage may show up under other windows. I believe `requestActivate()` tells WM a window requires attention. So I guess these 3 calls are necessary.

**Comment by vSLG** [comment]
The `.show` file is not working anymore, will fix it.

File: src/gui/Base/HTrayIcon.qml [hunk 12]
```diff
@@ -0,0 +1,57 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+
+import Qt.labs.platform 1.1
+import Qt.labs.folderlistmodel 2.12
+
+SystemTrayIcon {
+    property string iconPack: theme ? theme.icons.preferredPack : "thin"
+    property alias settingsFolder: showUpWatcher.folder
+
+    property var window
```
Comment (mirukana) [inline src/gui/Base/HTrayIcon.qml:10]: "Required" properties should be at the beginning, also you should use `ApplicationWindow` for the type instead of `var`

File: src/gui/Base/HTrayIcon.qml [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+        else if (reason !== SystemTrayIcon.Context)
+            window.visible ? window.hide() : window.restoreFromTray()
+    }
+
+    menu: Menu {
+        MenuItem {
+            text:
+                window.visible ?
+                "Minimize to tray" :
+                qsTr("Open ") + Qt.application.displayName
```
Comment (mirukana) [inline src/gui/Base/HTrayIcon.qml:45]: Use [`%x` parameters](https://doc.qt.io/qt-5/qtquick-internationalization.html#4-use-op-op-x-to-insert-parameters-into-a-string) for translatable strings, same for the Quit menu item

File: src/gui/Window.qml [hunk 14]
```diff
@@ -100,4 +111,9 @@ ApplicationWindow {
 
         Behavior on scale { HNumberAnimation { overshoot: 3; factor: 1.2 } }
     }
+
+    HTrayIcon {
```
Comment (mirukana) [inline src/gui/Window.qml:115]: The tray icon isn't a reusable component, it is specific to the Window, so it should be in the *gui* folder instead of *gui/Base*. The `H` prefix is used for component names that can conflict with standard ones (e.g. `Button` → `HButton`), so you can just name it `TrayIcon` in this case.

File: src/main.cpp [hunk 15]
```diff
@@ -184,19 +233,22 @@ int main(int argc, char *argv[]) {
...
-            qFatal(
+            qCritical(
                 "%s:%d:%d: %s",
                 e.url().toString().toStdString().c_str(),
                 e.line(),
                 e.column(),
                 e.description().toStdString().c_str()
             );
         }
+        qFatal("One or more errors have occurred, exiting.");
```
Comment (mirukana) [inline src/main.cpp:244]: The change here should be put in a separate commit

**Review by mirukana — CHANGES_REQUESTED** [review]
